### PR TITLE
fix(logger): force Logger to use local timezone when UTC flag is not set

### LINKED ISSUE #3154: Bug: Logger Timestamps Misaligned with Local Time
### Expected Behaviour

The logger should emit records using local time timestamp by default as per [docs](https://docs.powertools.aws.dev/lambda/python/latest/core/logger/#setting-timestamp-to-utc).

### Current Behaviour

logger emits records using `utc` timestamp when running in Australia/Sydney timezoneimage

### Code snippet

```python
from aws_lambda_powertools import Logger
from aws_lambda_powertools.event_handler import APIGatewayRestResolver
from aws_lambda_powertools.logging import correlation_paths

logger = Logger(service="APP")

app = APIGatewayRestResolver()


@app.get("/hello")
def hello():
    logger.info("Request from unknown received")
    return {"message": "hello unknown!"}


@logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_REST, log_event=True)
def lambda_handler(event, context):
    return app.resolve(event, context)
```

### Possible Solution

_No response_

### Steps to Reproduce

Run above lambda code snippet in **Sydney region (ap-southeast-2)**.

### Powertools for AWS Lambda (Python) version

latest

### AWS Lambda function runtime

3.9

### Packaging format used

Lambda Layers

### Debugging logs

_No response_

### PR DESCRIPTION
**Pull request by roger-zhangg**
**Issue number:** #3154

### Changes

Previously logging ignores TZ env value, this fix make logging follows local TZ value when UTC flag is not set 

### User experience

UX not affected

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: docs/core/logger.md [hunk 1]
```diff
@@ -39,7 +39,7 @@ Your Logger will include the following keys to your structured logging:
 | **level**: `str`           | `INFO`                                | Logging level                                                                                                                        |
 | **location**: `str`        | `collect.handler:1`                   | Source code location where statement was executed                                                                                    |
 | **message**: `Any`         | `Collecting payment`                  | Unserializable JSON values are casted as `str`                                                                                       |
-| **timestamp**: `str`       | `2021-05-03 10:20:19,650+0200`        | Timestamp with milliseconds, by default uses local timezone                                                                          |
+| **timestamp**: `str`       | `2021-05-03 10:20:19,650+0000`        | Timestamp with milliseconds, by default uses default lambda timezone (UTC)                                                                         |
```
Comment (leandrodamascena) [inline docs/core/logger.md]:
```suggestion
| **timestamp**: `str`       | `2021-05-03 10:20:19,650+0000`        | Timestamp with milliseconds, by default uses default AWS Lambda timezone (UTC)                                                                         |
```

File: docs/core/logger.md [hunk 2]
```diff
@@ -567,17 +567,24 @@ You can change the order of [standard Logger keys](#standard-structured-keys) or
     --8<-- "examples/logger/src/reordering_log_keys_output.json"
     ```
 
-#### Setting timestamp to UTC
+#### Setting timestamp to custom Timezone
 
-By default, this Logger and standard logging library emits records using local time timestamp. You can override this behavior via `utc` parameter:
+By default, this Logger and standard logging library emits records using default lambda time(UTC) timestamp. You can enforce this behavior via `utc` parameter.
```
Comment (leandrodamascena) [inline docs/core/logger.md]:
```suggestion
By default, this Logger and the standard logging library emit records with the default AWS Lambda timestamp in UTC.
```

I don't think we need to mention the enforce, this is the default behavior. To be honest, I think there is no use for the `utc` field anymore.

File: docs/core/logger.md [hunk 3]
```diff
@@ -567,17 +567,24 @@ You can change the order of [standard Logger keys](#standard-structured-keys) or
...
 
-=== "setting_utc_timestamp.py"
+However, if you want to use your preferred timezone to format your log instead, you can use the timezone environment variable `TZ`: Either set it as Lambda Environment Variable. Or setup this value directly in your Lambda Hander.
```
Comment (leandrodamascena) [inline docs/core/logger.md]:
```suggestion
However, if you want to use your preferred timezone to format your log instead, you can use the timezone environment variable `TZ`: Either set it as AWS Lambda environment variable, or setup this value directly in your Lambda function.
```

File: docs/core/logger.md [hunk 4]
```diff
@@ -567,17 +567,24 @@ You can change the order of [standard Logger keys](#standard-structured-keys) or
...
 
-    ```python hl_lines="6"
+???+ tip
+    `TZ` environment variable will be ingnored if `utc` is set to `True`
```
Comment (leandrodamascena) [inline docs/core/logger.md]:
```suggestion
    `TZ` environment variable will be ignored if `utc` is set to `True`
```

File: docs/core/logger.md [hunk 5]
```diff
@@ -567,17 +567,24 @@ You can change the order of [standard Logger keys](#standard-structured-keys) or
...
+
+=== "setting_custom_timezone.py"
+
+    ```python hl_lines="9 12"
     --8<-- "examples/logger/src/setting_utc_timestamp.py"
     ```
 
-=== "setting_utc_timestamp_output.json"
+    1.  if you set TZ in lambda runtime, time.tzset() need to be called. You don't need it when setting TZ in lambda environment variable
```
Comment (leandrodamascena) [inline docs/core/logger.md]:
```suggestion
    1.  if you set TZ in your Lambda function, `time.tzset()` need to be called. You don't need it when setting TZ in AWS Lambda environment variable
```

File: examples/logger/src/setting_utc_timestamp.py [hunk 6]
```diff
@@ -1,7 +1,13 @@
+import os
+import time
+
 from aws_lambda_powertools import Logger
 
-logger = Logger(service="payment")
-logger.info("Local time")
+logger = Logger(service="payment", utc=True)
+logger.info("utf time")
+
```
Comment (leandrodamascena) [inline examples/logger/src/setting_utc_timestamp.py]:
```suggestion
logger_in_utc = Logger(service="payment")
logger_in_utc.info("Logging with default AWS Lambda timezone: UTC time")

```

File: examples/logger/src/setting_utc_timestamp.py [hunk 7]
```diff
@@ -1,7 +1,13 @@
...
+os.environ["TZ"] = "US/Eastern"
+time.tzset()  # (1)!
 
-logger_in_utc = Logger(service="order", utc=True)
-logger_in_utc.info("GMT time zone")
+logger_in_utc = Logger(service="order")
+logger_in_utc.info("US eastern time")
```
Comment (leandrodamascena) [inline examples/logger/src/setting_utc_timestamp.py]:
```suggestion
logger = Logger(service="order")
logger.info("Logging with US Eastern timezone")
```

File: examples/logger/src/setting_utc_timestamp_output.json [hunk 8]
```diff
@@ -1,16 +1,16 @@
...
-        "level": "INFO",
-        "location": "<module>:4",
-        "message": "Local time",
-        "timestamp": "2022-06-24 11:39:49,421+0200",
-        "service": "payment"
+        "level":"INFO",
+        "location":"<module>:6",
+        "message":"utf time",
+        "timestamp":"2023-10-06 02:46:54,097+0000",
+        "service":"payment"
```
Comment (leandrodamascena) [inline examples/logger/src/setting_utc_timestamp_output.json]: After changing the example below I think we need to update this output.

Comment (leandrodamascena) [inline examples/logger/src/setting_utc_timestamp_output.json]: After changing the example below I think we need to update this output.

Comment (leandrodamascena) [inline examples/logger/src/setting_utc_timestamp_output.json]: After changing the example below I think we need to update this output.

File: examples/logger/src/setting_utc_timestamp_output.json [hunk 9]
```diff
@@ -1,16 +1,16 @@
...
-        "level": "INFO",
-        "location": "<module>:7",
-        "message": "GMT time zone",
-        "timestamp": "2022-06-24 09:39:49,421+0100",
-        "service": "order"
+        "level":"INFO",
+        "location":"<module>:12",
+        "message":"US eastern time",
+        "timestamp":"2023-10-05 22:46:54,097-0400",
+        "service":"order"
```
Comment (leandrodamascena) [inline examples/logger/src/setting_utc_timestamp_output.json]: same here.

File: package.json [hunk 10]
```diff
@@ -1,9 +1,6 @@
 {
   "name": "aws-lambda-powertools-python-e2e",
   "version": "1.0.0",
-  "devDependencies": {
-    "aws-cdk": "^2.99.1"
-  },
```
Comment (leandrodamascena) [inline package.json]: Is it right?

File: aws_lambda_powertools/logging/formatter.py [hunk 11]
```diff
@@ -144,6 +144,8 @@ def __init__(
 
         if self.utc:
             self.converter = time.gmtime
+        else:
+            self.converter = time.localtime
```
Comment (leandrodamascena) [inline aws_lambda_powertools/logging/formatter.py:150]: What you found here to simplify the solution was incredible! You always pay close attention to details. :clap:

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hi @roger-zhangg! Another PR with simple and extremely high-quality solutions. You're smashing it, man.

I left some comments we need to address before merging this!

File: package.json (hunk 10 above)
Comment (roger-zhangg) [reply package.json]: Interesting, I don't even know what caused this change. I'll revert

**Review by leandrodamascena — DISMISSED** [review]
@heitorlessa or @rubenfonseca, can you make a last review of this PR, please?

LGTM! Approving!

**Comment by roger-zhangg** [comment]
e2e test output

```bash
poetry run pytest tests/e2e/logger/test_logger.py
/local/home/ruojiazh/projects/aws-lambda-powertools-python/.venv/lib/python3.10/site-packages/pytest_benchmark/logger.py:46: PytestBenchmarkWarning: Benchmarks are automatically disabled because xdist plugin is active.Benchmarks cannot be performed reliably in a parallelized environment.
  warner(PytestBenchmarkWarning(text))
================================================================== test session starts ==================================================================
platform linux -- Python 3.10.2, pytest-7.4.2, pluggy-1.2.0 -- /local/home/ruojiazh/projects/aws-lambda-powertools-python/.venv/bin/python
cachedir: .pytest_cache
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /local/home/ruojiazh/projects/aws-lambda-powertools-python/tests/e2e
configfile: pytest.ini
plugins: typeguard-2.13.3, ddtrace-1.15.2, anyio-3.7.1, cov-4.1.0, xdist-3.3.1, mock-3.11.1, benchmark-4.0.0, socket-0.6.0, asyncio-0.21.1
asyncio: mode=strict
collected 13 items                                                                                                                                      

tests/e2e/logger/test_logger.py::test_basic_lambda_logs_visible PASSED                                                                            [  7%]
tests/e2e/logger/test_logger.py::test_lambda_tz[%z-US/Eastern-False] PASSED                                                                       [ 15%]
tests/e2e/logger/test_logger.py::test_lambda_tz[%z-US/Eastern-True] PASSED                                                                        [ 23%]
tests/e2e/logger/test_logger.py::test_lambda_tz[%z-UTC-False] PASSED                                                                              [ 30%]
tests/e2e/logger/test_logger.py::test_lambda_tz[%z-UTC-True] PASSED                                                                               [ 38%]
tests/e2e/logger/test_logger.py::test_lambda_tz[%z-Asia/Shanghai-False] PASSED                                                                    [ 46%]
tests/e2e/logger/test_logger.py::test_lambda_tz[%z-Asia/Shanghai-True] PASSED                                                                     [ 53%]
tests/e2e/logger/test_logger.py::test_lambda_tz[None-US/Eastern-False] PASSED                                                                     [ 61%]
tests/e2e/logger/test_logger.py::test_lambda_tz[None-US/Eastern-True] PASSED                                                                      [ 69%]
tests/e2e/logger/test_logger.py::test_lambda_tz[None-UTC-False] PASSED                                                                            [ 76%]
tests/e2e/logger/test_logger.py::test_lambda_tz[None-UTC-True] PASSED                                                                             [ 84%]
tests/e2e/logger/test_logger.py::test_lambda_tz[None-Asia/Shanghai-False] PASSED                                                                  [ 92%]
tests/e2e/logger/test_logger.py::test_lambda_tz[None-Asia/Shanghai-True] PASSED                                                                   [100%]

=================================================================== warnings summary ====================================================================
.venv/lib/python3.10/site-packages/ddtrace/internal/module.py:220
  /local/home/ruojiazh/projects/aws-lambda-powertools-python/.venv/lib/python3.10/site-packages/ddtrace/internal/module.py:220: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    self.loader.exec_module(module)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= 13 passed, 1 warning in 215.94s (0:03:35) =======================================================
(23-10-10 15:59:23) <0> [~/projects/aws-lambda-powertools-python]  
```

File: tests/e2e/logger/handlers/tz_handler.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+    logger = Logger(service=f"{utc}-{datefmt}-{tz}", utc=utc, datefmt=datefmt)
+    if logger.handlers[0].formatter.converter == time.localtime:
+        loggerType = "localtime_converter"
+    elif logger.handlers[0].formatter.converter == time.gmtime:
+        loggerType = "gmtime_converter"
+    else:
+        loggerType = "unknown"
+    logger.info(loggerType)
+    return "success"
```
Comment (heitorlessa) [inline tests/e2e/logger/handlers/tz_handler.py]: Add a comment to explain reasoning here. It's a bit hard to follow

File: tests/e2e/logger/test_logger.py [hunk 13]
```diff
@@ -36,3 +48,47 @@ def test_basic_lambda_logs_visible(basic_handler_fn, basic_handler_fn_arn):
...
     assert logs.have_keys(*LOGGER_LAMBDA_CONTEXT_KEYS) is True
+
+
+# test on combination of utc,tz,datefmt params
+@pytest.mark.xdist_group(name="logger")
+@pytest.mark.parametrize("utc", [False, True])
+@pytest.mark.parametrize("tz", ["US/Eastern", "UTC", "Asia/Shanghai"])
+@pytest.mark.parametrize("datefmt", ["%z", None])
+def test_lambda_tz(tz_handler_fn, tz_handler_fn_arn, utc, tz, datefmt):
+    # GIVEN
```
Comment (heitorlessa) [inline tests/e2e/logger/test_logger.py]: add details on GIVEN WHEN THEN. This helps self-document tests. Specially useful for parametrized tests that are testing more than one scenario

File: tests/e2e/logger/test_logger.py [hunk 14]
```diff
@@ -36,3 +48,47 @@ def test_basic_lambda_logs_visible(basic_handler_fn, basic_handler_fn_arn):
...
+    else:
+        # then use local time converter
+        assert result.message == "localtime_converter"
+
+    if utc:
+        assert result.timestamp[-5:] == "+0000"
+    else:
+        os.environ["TZ"] = tz
+        time.tzset()
+        assert result.timestamp[-5:] == time.strftime("%z", time.localtime())
```
Comment (heitorlessa) [inline tests/e2e/logger/test_logger.py]: this is typically a sign that we should have a test for local time and one for UTC.

If/Else and For loop are normally a sign that our tests are not straightforward enough - split this.

**Review by heitorlessa — CHANGES_REQUESTED** [review]
looks mostly good - E2E test should be split, and Lambda Handler needs comments.

A rule of thumb is that the minute you need If/Else, For/While Loops in a test, you need to break them. 

We run tests in parallel so there isn't a penalty for adding more granular E2E tests in this case.

**Review by leandrodamascena — APPROVED** [review]
LGTM!

**Review by heitorlessa — APPROVED** [review]
